File: ScalaKurs(Knabe)_J/src/ueb05/MathServiceImpl.java
package ueb05;

public class MathServiceImpl implements MathService {

	//---- recursive version
	public double integrate(final Function<Double> f, final double endX, final double incrementX) {
		return endX <= 0 ? 0 : incrementX * f.f(endX) + integrate(f, endX - incrementX, incrementX);
	}

	
	//---- tail-recursive version
//	public double integrate(final Function<Double> f, final double endX, final double incrementX) {
//		return integrate(f, endX - incrementX, incrementX, 0);
//	}
//	
//	private double integrate(final Function<Double> f, final double endX, final double incrementX, final double sum) {
//		if (endX <= 0) {
//			return sum; 
//		}
//		return integrate(f, endX - incrementX, incrementX, sum + incrementX * f.f(endX));
//	}
	
	
	//---- iterative version
//	public double integrate(final Function<Double> f, final double endX, final double incrementX) {
//		double sum = 0;
//		for (double x = incrementX; x <= endX; x += incrementX) {
//			sum += incrementX * f.f(x);
//		}
//		return sum;
//	}
	
	
	public Function<Double> integral(final Function<Double> f, final double incrementX) {
		return new Function<Double>() {
			public Double f(final Double x) {
				return integrate(f, x, incrementX);
			}
		};
	}

}

File: ScalaKurs(Knabe)_S/src/methfunc/MathServiceImpl.scala
package methfunc

import scala.annotation.tailrec

/** A Service for numerically integrating functions of double floating point numbers.*/
class MathServiceImpl {
  
  /** Computes the area between the X-axis and the graph of the function f from 0 to endX 
   * by summing up the rectangles beneath the graph with a width of incrementX. */
  def integrate(f: Double => Double, endX: Double, incrementX: Double): Double = {
	if (endX <= 0) 0 
	else incrementX * f(endX) + integrate(f, endX - incrementX, incrementX)
  }
	
//  //---- tail-recursive version
//  def integrate(f: Double => Double, endX: Double, incrementX: Double): Double = {
//	integrate(f, endX - incrementX, incrementX, 0)
//  }
//  
//  @tailrec private def integrate(f: Double => Double, endX: Double, incrementX: Double, sum: Double): Double = {
//    if (endX <= 0) sum 
//    else integrate(f, endX - incrementX, incrementX, sum + incrementX * f(endX))
//  }
    
  /** Returns a function, which is the integration function for f, computed by X-increments of incrementX. */
  val integral = (f: Double => Double, incrementX: Double) => integrate(f, _: Double, incrementX)
  
//  val integral: (Double => Double, Double) => (Double => Double) = (f, incrementX) => {
//	integrate(f, _, incrementX)
//  }

}